File: apps/api/src/Erudoza.Domain/Study/MissingWordAnswers.cs
namespace Erudoza.Domain.Study;

public sealed record MissingWordAnswer(int Index, string Text);

public sealed record MissingWordResult(int Index, bool IsCorrect, string Expected);

public sealed record MissingWordAnswerPayload(string Format, IReadOnlyList<MissingWordAnswer> Answers, IReadOnlyList<MissingWordResult> Results)
{
    public const string CurrentFormat = "missing-words-slots/v1";
    private static readonly System.Text.Json.JsonSerializerOptions Json = new(System.Text.Json.JsonSerializerDefaults.Web);
    public string Serialize() => System.Text.Json.JsonSerializer.Serialize(this, Json);
    public static MissingWordAnswerPayload? Read(string? json) => json is null ? null : System.Text.Json.JsonSerializer.Deserialize<MissingWordAnswerPayload>(json, Json);
}

public sealed record MissingWordEvaluation(bool IsCorrect, IReadOnlyList<MissingWordResult> Results);

public static class MissingWordAnswers
{
    public static MissingWordEvaluation Evaluate(
        IReadOnlyList<MissingWordsToken> tokens,
        IReadOnlyList<MissingWordAnswer> answers)
    {
        if (tokens is null || answers is null)
        {
            throw InvalidAnswers();
        }

        var tokenIndices = new HashSet<int>();
        foreach (var token in tokens)
        {
            if (!tokenIndices.Add(token.Index))
            {
                throw InvalidAnswers();
            }
        }

        var hidden = tokens.Where(token => token.Hidden).ToList();
        var hiddenIndices = hidden.Select(token => token.Index).ToHashSet();
        var submitted = new Dictionary<int, string>();
        foreach (var answer in answers)
        {
            if (answer is null || answer.Text is null || !hiddenIndices.Contains(answer.Index) ||
                !submitted.TryAdd(answer.Index, answer.Text))
            {
                throw InvalidAnswers();
            }
        }

        if (submitted.Count != hidden.Count)
        {
            throw InvalidAnswers();
        }

        var results = hidden.Select(token =>
        {
            var text = submitted[token.Index];
            var normalizedExpected = TextNormalization.Normalize(token.Text, NormalizationProfile.ExactText);
            var normalizedSubmitted = TextNormalization.Normalize(text, NormalizationProfile.ExactText);
            var isCorrect = normalizedExpected.Length > 0
                ? normalizedSubmitted.Length > 0 && normalizedSubmitted == normalizedExpected
                : text.Trim().Length > 0 && string.Equals(text.Trim(), token.Text.Trim(), StringComparison.Ordinal);
            return new MissingWordResult(token.Index, isCorrect, token.Text);
        }).ToList();

        return new MissingWordEvaluation(results.All(result => result.IsCorrect), results);
    }

    private static DomainException InvalidAnswers() => new("Invalid missing-word answers.");
}

File: apps/web/worker/native/study/missing-word-answers.ts
import { normalizeText, type Token } from './engine';

export interface MissingWordAnswer { index: number; text: string }
export interface MissingWordResult { index: number; isCorrect: boolean; expected: string }

export function evaluateMissingWordAnswers(tokens: Token[], answers: MissingWordAnswer[]): {
  isCorrect: boolean;
  results: MissingWordResult[];
  score: number;
  evaluationCode: 'ExactMatch' | 'PartialMatch' | 'Incorrect';
} {
  const tokenIndices = new Set<number>();
  for (const token of tokens) {
    if (!Number.isInteger(token.index) || tokenIndices.has(token.index)) throw new Error('Invalid missing-word answers.');
    tokenIndices.add(token.index);
  }

  const hidden = tokens.filter(token => token.hidden);
  const hiddenIndices = new Set(hidden.map(token => token.index));
  const submitted = new Map<number, string>();
  if (!Array.isArray(answers)) throw new Error('Invalid missing-word answers.');
  for (const answer of answers) {
    if (!answer || !Number.isInteger(answer.index) || typeof answer.text !== 'string' ||
      !hiddenIndices.has(answer.index) || submitted.has(answer.index)) {
      throw new Error('Invalid missing-word answers.');
    }
    submitted.set(answer.index, answer.text);
  }
  if (submitted.size !== hidden.length) throw new Error('Invalid missing-word answers.');

  const results = hidden.map(token => {
    const text = submitted.get(token.index)!;
    const normalizedExpected = normalizeText(token.text);
    const isCorrect = normalizedExpected.length > 0
      ? normalizeText(text).length > 0 && normalizeText(text) === normalizedExpected
      : text.trim().length > 0 && text.trim() === token.text.trim();
    return { index: token.index, isCorrect, expected: token.text };
  });
  const isCorrect = results.every(result => result.isCorrect);
  const score = results.length ? Math.round(results.filter(result => result.isCorrect).length / results.length * 100) : 0;
  return { isCorrect, results, score, evaluationCode: isCorrect ? 'ExactMatch' : score >= 50 ? 'PartialMatch' : 'Incorrect' };
}
